File: cchess/lib/core/chess_engine/piece.dart
import 'package:equatable/equatable.dart';

import '../constants/piece_constants.dart';

/// Immutable Xiangqi piece: a type + color pair.
class Piece extends Equatable {
  final PieceType type;
  final PieceColor color;

  const Piece(this.type, this.color);

  // Convenient factories.
  static const Piece redGeneral = Piece(PieceType.general, PieceColor.red);
  static const Piece redAdvisor = Piece(PieceType.advisor, PieceColor.red);
  static const Piece redElephant = Piece(PieceType.elephant, PieceColor.red);
  static const Piece redHorse = Piece(PieceType.horse, PieceColor.red);
  static const Piece redChariot = Piece(PieceType.chariot, PieceColor.red);
  static const Piece redCannon = Piece(PieceType.cannon, PieceColor.red);
  static const Piece redSoldier = Piece(PieceType.soldier, PieceColor.red);

  static const Piece blackGeneral = Piece(PieceType.general, PieceColor.black);
  static const Piece blackAdvisor = Piece(PieceType.advisor, PieceColor.black);
  static const Piece blackElephant = Piece(PieceType.elephant, PieceColor.black);
  static const Piece blackHorse = Piece(PieceType.horse, PieceColor.black);
  static const Piece blackChariot = Piece(PieceType.chariot, PieceColor.black);
  static const Piece blackCannon = Piece(PieceType.cannon, PieceColor.black);
  static const Piece blackSoldier = Piece(PieceType.soldier, PieceColor.black);

  bool get isRed => color == PieceColor.red;
  bool get isBlack => color == PieceColor.black;

  String get fenChar => type.fenLetter(color);
  String get hanChar => type.hanChar(color);

  @override
  String toString() => fenChar;

  @override
  List<Object?> get props => [type, color];
}

File: cchess-backend/src/engine/piece.ts
// Port of cchess/lib/core/constants/piece_constants.dart + piece.dart
// Convention: Red is the bottom side (row 9), Black is the top (row 0).

export enum PieceColor {
  Red = 'red',
  Black = 'black',
}

export function oppositeColor(c: PieceColor): PieceColor {
  return c === PieceColor.Red ? PieceColor.Black : PieceColor.Red;
}

export enum PieceType {
  General = 'general',
  Advisor = 'advisor',
  Elephant = 'elephant',
  Horse = 'horse',
  Chariot = 'chariot',
  Cannon = 'cannon',
  Soldier = 'soldier',
}

export class Piece {
  constructor(
    public readonly type: PieceType,
    public readonly color: PieceColor,
  ) {}

  get isRed(): boolean {
    return this.color === PieceColor.Red;
  }

  fenChar(): string {
    return PieceTypeFenLetter(this.type, this.color);
  }

  equals(other: Piece): boolean {
    return this.type === other.type && this.color === other.color;
  }

  toString(): string {
    return this.fenChar();
  }
}

export function PieceTypeFenLetter(t: PieceType, c: PieceColor): string {
  let letter: string;
  switch (t) {
    case PieceType.General: letter = 'k'; break;
    case PieceType.Advisor: letter = 'a'; break;
    case PieceType.Elephant: letter = 'b'; break;
    case PieceType.Horse: letter = 'n'; break;
    case PieceType.Chariot: letter = 'r'; break;
    case PieceType.Cannon: letter = 'c'; break;
    case PieceType.Soldier: letter = 'p'; break;
  }
  return c === PieceColor.Red ? letter.toUpperCase() : letter;
}

/// Parse a FEN letter to [type, color]. Returns null for empty squares.
export function fromFenLetter(letter: string): [PieceType, PieceColor] | null {
  if (letter.length === 0) return null;
  const isRed = letter === letter.toUpperCase();
  const color = isRed ? PieceColor.Red : PieceColor.Black;
  switch (letter.toLowerCase()) {
    case 'k': return [PieceType.General, color];
    case 'a': return [PieceType.Advisor, color];
    case 'b':
    case 'e': return [PieceType.Elephant, color];
    case 'n':
    case 'h': return [PieceType.Horse, color];
    case 'r': return [PieceType.Chariot, color];
    case 'c': return [PieceType.Cannon, color];
    case 'p': return [PieceType.Soldier, color];
    default: return null;
  }
}

export const INITIAL_FEN =
  'rnbakabnr/9/1c5c1/p1p1p1p1p/9/9/P1P1P1P1P/1C5C1/9/RNBAKABNR w - - 0 1';
